falcon/media/json.py

```diff
@@ -148,55 +153,73 @@ def default(self, obj):
         loads (func): Function to use when deserializing JSON requests.
     """
 
-    def __init__(self, dumps=None, loads=None):
+    def __init__(
+        self,
+        dumps: Optional[Callable[[Any], Union[str, bytes]]] = None,
+        loads: Optional[Callable[[str], Any]] = None,
+    ) -> None:
         self._dumps = dumps or partial(json.dumps, ensure_ascii=False)
         self._loads = loads or json.loads
 
         # PERF(kgriffs): Test dumps once up front so we can set the
         #     proper serialize implementation.
         result = self._dumps({'message': 'Hello World'})
         if isinstance(result, str):
-            self.serialize = self._serialize_s
-            self.serialize_async = self._serialize_async_s
+            self.serialize = self._serialize_s  # type: ignore[method-assign]
+            self.serialize_async = self._serialize_async_s  # type: ignore[method-assign]
         else:
-            self.serialize = self._serialize_b
-            self.serialize_async = self._serialize_async_b
+            self.serialize = self._serialize_b  # type: ignore[method-assign]
+            self.serialize_async = self._serialize_async_b  # type: ignore[method-assign]
 
         # NOTE(kgriffs): To be safe, only enable the optimized protocol when
         #   not subclassed.
         if type(self) is JSONHandler:
             self._serialize_sync = self.serialize
             self._deserialize_sync = self._deserialize
 
-    def _deserialize(self, data):
+    def _deserialize(self, data: bytes) -> Any:
         if not data:
             raise errors.MediaNotFoundError('JSON')
         try:
             return self._loads(data.decode())
         except ValueError as err:
             raise errors.MediaMalformedError('JSON') from err
 
-    def deserialize(self, stream, content_type, content_length):
+    def deserialize(
+        self,
+        stream: ReadableIO,
+        content_type: Optional[str],
+        content_length: Optional[int],
+    ) -> Any:
         return self._deserialize(stream.read())
 
-    async def deserialize_async(self, stream, content_type, content_length):
+    async def deserialize_async(
+        self,
+        stream: AsyncReadableIO,
+        content_type: Optional[str],
+        content_length: Optional[int],
+    ) -> Any:
         return self._deserialize(await stream.read())
 
     # NOTE(kgriffs): Make content_type a kwarg to support the
     #   Request.render_body() shortcut optimization.
-    def _serialize_s(self, media, content_type=None) -> bytes:
-        return self._dumps(media).encode()
+    def _serialize_s(self, media: Any, content_type: Optional[str] = None) -> bytes:
```

Comment:
I suppose `media` could be of type `JSONSerializable` that I introduced to type SSE. But probably it causes more trouble than benefits...

Reply:
yes, that's my impression.
using Any has the advantage that you can do `foo: MyTypedDict = req.get_media()` and the type checker is happy

Reply:
I added a point to the cleanup issue to weigh pros vs cons of this one.